api/String.h

```diff
@@ -81,7 +81,7 @@ class String
 	explicit String(unsigned long, unsigned char base=10);
 	explicit String(float, unsigned char decimalPlaces=2);
 	explicit String(double, unsigned char decimalPlaces=2);
-	~String(void);
+	virtual ~String(void);
```

Comment:
Why do you think that `String `warrants a `virtual `destructor? Is there any `virtual `function in the class? On a quick glance I could not find one.

Reply:
This question answer is as simple as understanding if we want to allow the user to extend the `arduino::String` class

Reply:
> This question answer is as simple as understanding if we want to allow the user to extend the `arduino::String` class

Undefined behaviour is related to having any virtual function, not to inheritance.
So currently, we do need to add a virtual destructor to `Print` because it has a virtual functions.

Thus, a virtual destructor is not currently a necessity for `String`. I can inherit from `String`, ex. `MyString` and use `MyString` all along my code, without any undefined behaviour.

However, you might want to add a virtual destructor to `String` because you expect users to add virtual functions to their `MyString`. But that's more about securing the code about expected usages than fixing the current code.

Reply:
I can have memory leaks if in `MyString` class I allocate something that needs to be deleted, I create a proper destructor, I reference it with a pointer to `String` then I delete it, the derived class destructor is not called. 

The answer, in my opinion, is not strictly limited to "`String` doesn't have virtual methods", because I could want some features of `String` class and extend them with my custom features. I think you can implement a solution in other better ways, like by embedding `String` instead. Thus my question: do we want to allow the user to extend `String` class?